### ISSUE TITLE
`gem update --system` fails with "wrong number of arguments (given 4, expected 1)"

### ISSUE THREAD

**Issue by dmolesUC** [jruby/jruby#7064]
**Environment Information**

- JRuby version 9.2.9.0; `jruby -v` output:
   ```none
   jruby 9.2.9.0 (2.5.7) 2019-10-30 458ad3e OpenJDK 64-Bit Server VM 17.0.1+12-LTS on 17.0.1+12-LTS +jit [darwin-aarch64]
   ```
- OS: macOS Monterey 12.2; `uname -a` output:
  ```none
  Darwin ArtMachine-2.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64
  ```
- Shell: zsh 5.8
- Ruby version manager: RVM 1.29.12-next
- Java versions tested:
  - 1.8.0_311 (Oracle / Intel)
  - 17.0.1 (OpenJDK Zulu / Apple Silicon)

**Steps to reproduce**

On a system with JRuby 9.2.9.0 installed with RVM:

- Clone the [ruby-marc](git@github.com:ruby-marc/ruby-marc.git) repository
- Check out commit [335a020](https://github.com/ruby-marc/ruby-marc/commit/335a02065136440c8fb57c8f18e99dcd6e23791a)
- Select JRuby with `rvm use jruby`
- run `bundle install`

**Expected Behavior**

1. `bundle install` succeeds without major warnings
2. when the above proves not to be the case (see below):
   - `gem update --system` succeeds

**Actual Behavior**

1. `bundle install` shows the following warning:

   ```none
   Your RubyGems version (2.7.10)) has a bug that prevents `required_ruby_version`
   from working for Bundler. Any scripts that use `gem install bundler` will break
   as soon as Bundler drops support for your Ruby version. Please upgrade RubyGems
   to avoid future breakage and silence this warning by running
   `gem update --system 3.2.3`
   ```

2. `gem update --system` fails with

   ```none
   Updating rubygems-update
   ERROR:  While executing gem ... (ArgumentError)
       wrong number of arguments (given 4, expected 1)
   ```

   Full stack trace:

   ```none
   $ jruby -S gem update --backtrace --system
   Updating rubygems-update
   ERROR:  While executing gem ... (ArgumentError)
	wrong number of arguments (given 4, expected 1)
	/Users/david/.rvm/gems/jruby-9.2.9.0/gems/psych-4.0.3-java/lib/psych.rb:31:in `safe_load'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/safe_yaml.rb:31:in `safe_load'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:504:in `block in read_checksums'
	org/jruby/ext/zlib/RubyGzipFile.java:97:in `wrap'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:503:in `block in read_checksums'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package/tar_reader.rb:116:in `seek'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:502:in `read_checksums'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:555:in `block in verify'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package/tar_reader.rb:29:in `new'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:554:in `block in verify'
	org/jruby/RubyIO.java:1160:in `open'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package/file_source.rb:30:in `with_read_io'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:553:in `verify'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/package.rb:534:in `spec'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/installer.rb:267:in `spec'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/dependency_installer.rb:412:in `block in install'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/request_set.rb:169:in `block in install'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/resolver/specification.rb:95:in `install'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/request_set.rb:168:in `block in install'
	org/jruby/RubyArray.java:1814:in `each'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/request_set.rb:156:in `install'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/dependency_installer.rb:411:in `install'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/commands/update_command.rb:213:in `update_gem'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/commands/update_command.rb:241:in `update_rubygems'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/commands/update_command.rb:90:in `execute'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/command.rb:313:in `invoke_with_build_args'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/command_manager.rb:173:in `process_args'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/command_manager.rb:143:in `run'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/lib/ruby/stdlib/rubygems/gem_runner.rb:59:in `run'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/bin/jgem:21:in `<main>'
	org/jruby/RubyKernel.java:1013:in `load'
	/Users/david/.rvm/rubies/jruby-9.2.9.0/bin/gem:4:in `<main>'
   ```

**Comment by headius** [comment]
This is an incompatibility between RubyGems 2.7.10 (shipped with JRuby 9.2.9.0) and Psych 4.0. Versions of Psych after 3.1.0.pre1 shifted this method from a four-argument form to a single argument plus three keyword arguments.

This is the code in a more recent RubyGems (such as the 3.1.6 version shipped :

```
def self.safe_load(input)
  if Gem::Version.new(Psych::VERSION) >= Gem::Version.new('3.1.0.pre1')
    ::YAML.safe_load(input, permitted_classes: PERMITTED_CLASSES, permitted_symbols: PERMITTED_SYMBOLS, aliases: true)
  else
    ::YAML.safe_load(input, PERMITTED_CLASSES, PERMITTED_SYMBOLS, true)
  end
end
```

But this is the code in 2.7.10:

```
def self.safe_load input
  ::YAML.safe_load(input, WHITELISTED_CLASSES, WHITELISTED_SYMBOLS, true)
end
```

If you uninstall the newer Psych versions you should be able to `update --system` and then reinstall Psych.

cc @deivid-rodriguez @tenderlove

**Comment by deivid-rodriguez** [comment]
Yes, that's exactly right. Thinking whether it might be worth to include some hint about this with that deprecation message?

**Comment by headius** [comment]
@deivid-rodriguez Unfortunately I'm not sure it would help... The situation here was that psych was updated before RubyGems, so the user was stuck with an broken setup. I think this could only be avoided by adding back to psych the four argument version and warning if those last three arguments are used. I'm guessing there was no transitional period where both the four argument version and the keyword version were working?

**Comment by deivid-rodriguez** [comment]
I was just thinking of saying something like

> If `gem update --system` fails with an "invalid number of arguments" error, make sure you uninstall psych 4 before attempting the upgrade, and you can install it again _after_ the successful upgrade.

Psych did this on a major version I think, which seems fine. Problem was RubyGems depended on the previous behavior.

**Comment by headius** [comment]
Regarding the change in psych, this is why breaking changes should keep the old behavior, with deprecation, for at least a few minor or one major version. There need to be a transitional period.

> I was just thinking of saying something like

Unfortunately the error occurs before the update gem is ever installed, so there's no way to detect this situation. Once the newer psych is installed, the older RubyGems is broken until it is removed. Perhaps we just need to SEO and get some visibility for this situation.

**Comment by deivid-rodriguez** [comment]
> Regarding the change in psych, this is why breaking changes should keep the old behavior, with deprecation, for at least a few minor or one major version. There need to be a transitional period.

Nothing to do about this now, I only noticed this when someone first reported it to me.

> Unfortunately the error occurs before the update gem is ever installed, so there's no way to detect this situation. Once the newer psych is installed, the older RubyGems is broken until it is removed. Perhaps we just need to SEO and get some visibility for this situation.

If I understand correctly, OP run into this issue while trying to fix a Bundler warning that suggests `gem update --system`, I'm thinking of enhancing this Bundler warning with information about this issue. We could even detect the installed version of psych when printing the warning, and only show it if it's Psych 4.

**Comment by dmolesUC** [comment]
I’m new to JRuby, but now that it’s happened I’m pretty sure I’ve been bit by this particular Psych/RubyGems issue on MRI. Not lately, I presume because most of our projects have moved to newer versions of Ruby and RubyGems. Until JRuby does that, it might be helpful to have something in the docs suggesting one consider running `gem update --system` *before* installing any gems.

**Comment by dmolesUC** [comment]
And yes, it was `bundle install` that pulled in Psych 4, as a transitive dependency of an open dependency on `rdoc`.

**Comment by headius** [comment]
> Until JRuby does that

The version of JRuby you reported this against was 9.2.9.0, which is over 2 years old (October 2019). There have been a dozen releases of 9.2 since then, plus four releases of 9.3. current versions of each ship with newer RubyGems that do not have this problem.

> it was bundle install that pulled in Psych 4

So it was not bundler itself that pulled in psych; it was the dependencies bundler installed for you. Was this a current version of bundler? I ask because if we add a better warning or better detection to bundler, it will only be useful if the newer version of bundler is installed prior to installing the newer version of psych. If you're set up here was using an older bundler you would never get such a warning.

**Comment by headius** [comment]
Basically it is a bit of a chicken or egg situation: if you have an older bundler that installs a new psych, there's never any chance to intervene and warn the user that the new site will break RubyGems. At that point it's impossible to install a new bundler.

Perhaps a new version of psych should be released that warns if it is being installed by an old version of RubyGems? Isn't there a way to specify a minimum version in the spec?

**Comment by deivid-rodriguez** [comment]
Interesting idea! And yes, this is so chicken and egg 😆. There's a way to specify a minimum version but it wouldn't help here because Bundler would still fallback to the previous compatible version, which would still be 4.x. An install time warning could help more, but it might be overkill. Overall, this should get better with time and I guess the best course of action is to apply the ostrich algorithm :)

**Comment by dmolesUC** [comment]
@headius Yes, 9.2.9.0 was just what I got with `rvm install jruby`, but I should know by now not to assume that RVM's default for a given platform is current (or even that its list of known versions is current).

@deivid-rodriguez Agreed, it should solve itself eventually. And if this issue had existed before I posted it, I would have found this discussion with the explanation and workarounds, so maybe this is good enough. 🙃
